internal/modules/modules.go

```diff
@@ -96,16 +98,214 @@ type ModuleResolveOptions struct {
+
+	log.WithFields(logrus.Fields{
+		"module":  item.importPath,
+		"version": version.GitRef(),
+	}).Debug("resolved module")
+	return nil
+}
+
+// workList is a list of modules to resolve
+type workList struct {
+	tasks []resolveModule
+	// replacements replace the URL for a module's
 	// provided import path.
-	strReplacements := make(map[string]string)
+	replacements map[string]string
 
 	// resolved contains the current modules that have been selected and is used
 	// to track previous resolutions/constraints for re-resolving modules.
-	resolved := make(map[string]*resolvedModule)
+	resolved map[string]*resolvedModule
+
+	// mu protects all the collections in this struct
+	mu  sync.Mutex
+	log logrus.FieldLogger
+}
+
+// workItem is a single item in the work list
+type workItem struct {
+	importPath           string
+	inProgressResolution *resolvedModule
+	spec                 resolveModule
+	uri                  string
+}
+
+// pop removes and returns the first item in the work list
+func (list *workList) pop() *workItem {
+	list.mu.Lock()
+	defer list.mu.Unlock()
+
+	if len(list.tasks) == 0 {
+		return nil
+	}
+	resolv := list.tasks[0]
+	list.tasks = list.tasks[1:]
+
+	if resolv.conf.Version != "" {
+		if _, err := semver.NewConstraint(resolv.conf.Version); err != nil {
+			// Attempt to resolve as a branch, which essentially is a channel.
+			// This is a bit of a hack, ideally we'll consolidate this logic when
+			// channels are ripped out of stencil later.
+			resolv.conf.Channel = resolv.conf.Version
+			resolv.conf.Version = ""
+		}
+	}
+
+	// check if it has already been resolved
+	importPath := resolv.conf.Name
+	if _, ok := list.resolved[importPath]; !ok {
+		list.resolved[importPath] = &resolvedModule{Module: &Module{}}
+	}
+	rm := list.resolved[importPath]
+
+	// if the module has already been resolved and is marked as
+	// "dontResolve", then re-use it.
+	if rm.dontResolve {
+		// this module has already been resolved and should always be used
+		list.log.WithFields(logrus.Fields{
+			"module": importPath,
+		}).Debug("Using in-memory module")
+		return &workItem{importPath: importPath, inProgressResolution: rm, spec: resolv}
+	}
+
+	// log the resolution attempt
+	rm.history = append(rm.history, resolution{
+		constraint:   resolv.conf.Version,
+		channel:      resolv.conf.Channel,
+		parentModule: resolv.parent,
+	})
+	list.log.WithFields(logrus.Fields{
+		"module": importPath,
+		"parent": resolv.parent,
+	}).Debug("resolving module")
+
+	uri := "https://" + importPath
+	// use a different url for the module if it's been replaced
+	if _, ok := list.replacements[importPath]; ok {
+		uri = list.replacements[importPath]
+	}
+
+	return &workItem{importPath: importPath, inProgressResolution: rm, spec: resolv, uri: uri}
+}
+
+// push adds a module to the work list
+func (list *workList) push(task *resolveModule) {
+	list.mu.Lock()
+	defer list.mu.Unlock()
+
+	list.tasks = append(list.tasks, *task)
```

Comment:
why deref as opposed to just queuing the pointers?

Reply:
Could do. Aren't slices already pointers,

Reply:
They're reference types, but that is just the container. You're still copying the entire `task` struct into the container (the slice) as opposed to just the reference here though

Reply:
true, i can make it a slice of pointers if you care. seemed low impact to me

Reply:
done.